### ISSUE TITLE
stdout pipe broken after container checkpoint/restore twice

### ISSUE THREAD

**Issue by carlhoerberg** [containers/podman#11911]
/kind bug

**Description**

After a checkpoint restore twice applications can't write to stdout.

**Steps to reproduce the issue:**

1. `sudo -i`
2. `podman run -d alpine sh -c "while true; do date; sleep 1; done"`
3. `podman container checkpoint --latest && podman container restore --latest`
4. `podman container checkpoint --latest && podman container restore --latest` 
5. `podman logs -f --latest`

**Describe the results you received:**

The container logs until checkpointed, never resumes after restore. 

`strace` show:

```
[pid 15967] 12:02:03 writev(1, [{iov_base="Tue Oct 12 12:02:03 UTC 2021", iov_len=28}, {iov_base="\n", iov_len=1}], 2) = -1 EPIPE (Broken pipe)
[pid 15967] 12:02:03 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=466, si_uid=0} ---
```

**Describe the results you expected:**

Expect STDOUT pipe not to be broken. 

**Additional information you deem important (e.g. issue happens only occasionally):**

Happens every time

**Output of `podman version`:**

```
Version:      3.4.0
API Version:  3.4.0
Go Version:   go1.16.8
Built:        Thu Sep 30 19:40:21 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.30-2.fc34.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.30, commit: '
  cpus: 2
  distribution:
    distribution: fedora
    variant: cloud
    version: "34"
  eventLogger: journald
  hostname: ip-172-31-6-203.ec2.internal
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.11.12-300.fc34.x86_64
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 1072766976
  memTotal: 2032918528
  ociRuntime:
    name: crun
    package: crun-1.2-1.fc34.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.2
      commit: 4f6c8e0583c679bfee6a899c05ac6b916022561b
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc34.x86_64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.0
  swapFree: 0
  swapTotal: 0
  uptime: 27h 28m 11.9s (Approximately 1.12 days)
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 1
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.4.0
  Built: 1633030821
  BuiltTime: Thu Sep 30 19:40:21 2021
  GitCommit: ""
  GoVersion: go1.16.8
  OsArch: linux/amd64
  Version: 3.4.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-3.4.0-1.fc34.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Fedora 34 (Cloud Edition), AWS

**Comment by carlhoerberg** [comment]
Not reproducible in podman `3.4.0`

**Comment by edsantiago** [comment]
Good thinking. criu is 3.16-2.fc34 on my system. But I `dnf downgrade`d to 3.15-3.fc34 and still can't reproduce it (using podman @ main, though; I didn't take the time to rebuild 3.3.1).

**Comment by carlhoerberg** [comment]
No, now I replicated it with podman 3.4.0 and criu 3.16, when restoring an exported checkpoint to another machine. 

When `strace`ing I see that the stdout pipe is boken: 

```
[pid 11635] 11:22:19 writev(1, [{iov_base="connected: 158.174.184.124:38434", iov_len=32}, {iov_base="\n", iov_len=1}], 2) = -1 EPIPE (Broken pipe)
```

**Comment by carlhoerberg** [comment]
oh, and `--tcp-established` is required too to replicate it. I've updated the description.

**Comment by carlhoerberg** [comment]
If checkpoint is run with `--keep`, the first time `dump.log` has:
```
(00.010302) 20332 fdinfo 0: pos:                0 flags:           100002/0
(00.010320) Dumping path for 0 fd via self 18 [/dev/null]
(00.010363) 20332 fdinfo 1: pos:                0 flags:                1/0
(00.010369) Dumping pipe 19 with id 0x4 pipe_id 0x13990
(00.010385) 20332 fdinfo 2: pos:                0 flags:                1/0
(00.010388) Dumping pipe 20 with id 0x5 pipe_id 0x13991
```

The second checkpoint dump.log has: 
```
(00.010385) 20601 fdinfo 0: pos:                0 flags:           100002/0
(00.010402) Dumping path for 0 fd via self 18 [/dev/null]
(00.010449) 20601 fdinfo 1: pos:                0 flags:           100001/0
(00.010455) Dumping pipe 19 with id 0x4 pipe_id 0x13bb1
(00.010469) 20601 fdinfo 2: pos:                0 flags:           100001/0
(00.010472) Dumping pipe 20 with id 0x5 pipe_id 0x13bb2

```

**Comment by mheon** [comment]
@adrianreber PTAL

**Comment by carlhoerberg** [comment]
Thank you! Yes, running with `--runtime runc` fixes it!

**Comment by adrianreber** [comment]
According to the changelog it should be fixed with 1.2. @giuseppe can you confirm that crun 1.2 should fix this? Maybe there is one more bug in crun.
